### PR TITLE
fix(router): update scope_key_builder when srds config is updated

### PR DESCRIPTION
**Pull request by tony612**
Commit Message: fix(router): update scope_key_builder when srds config is updated
Additional Description:

When SRDS is used and `scope_key_builder` in HCM is updated, the `ScopedConfigImpl`'s scope_key_builder is not updated.

I found the reason is when ConfigProvider is created in https://github.com/envoyproxy/envoy/blob/aac7cc259417a0c1d5b19b91bd34ac3800123a02/source/common/router/scoped_rds.cc#L619 cache is used in `getSubscription` and the cache key is [`scoped_rds`](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto#envoy-v3-api-field-extensions-filters-network-http-connection-manager-v3-scopedroutes-scoped-rds). So if only `scope_key_builder` is updated when HCM is updated via LDS, the previous `ScopedRdsConfigSubscription` is used directly, its `scope_key_builder` is not updated and `ScopedConfigImpl`'s scope_key_builder is not updated in worker threads either.

~My solution is adding a method `ScopedRdsConfigSubscription::updateScopeKeyBuilder` to update scope_key_builder in createXdsConfigProvider. This may be redundant in the first time, but it's seems okay for now.~

UPDATE: Another problem is when two HCM use the same SRDS xds config, they will share the same scope key builder. So I separated the scope key builder from `ScopedRdsConfigSubscription` to HCM to fix this problem.

As the unit test shows, without this fix, the new test will fail:
```
[----------] 1 test from ScopedRdsTest
[ RUN      ] ScopedRdsTest.OnScopeKeyBuilderUpdate
external/envoy/test/common/router/scoped_rds_test.cc:1781: Failure
Value of: config
Expected: isn't NULL
  Actual: 16-byte object <00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00>
```

I also added an integration test to test the case that two HCMs share the same SRDS xds config.

Risk Level: Low
Testing: unit test
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/extensions/listener_managers/listener_manager/lds_api.cc [hunk 1]
```diff
@@ -46,6 +47,7 @@ void LdsApiImpl::onConfigUpdate(const std::vector<Config::DecodedResourceRef>& a
   if (cm_.adsMux()) {
     const std::vector<std::string> paused_xds_types{
         Config::getTypeUrl<envoy::config::route::v3::RouteConfiguration>(),
+        Config::getTypeUrl<envoy::config::route::v3::ScopedRouteConfiguration>(),
```
Comment (tony612) [inline source/extensions/listener_managers/listener_manager/lds_api.cc:50]: This is not related to the issue. But I think we should pause SRDS when LDS is updated.

**Comment by kyessenov** [comment]
I'm not super familiar with SRDS, but I have a general question: shouldn't the key be calculated separately from the subscription? I wonder if we have two HCMs pointing to the same SRDS config source with different scope keys, what happens then?

File: source/extensions/listener_managers/listener_manager/lds_api.cc (hunk 1 above)
Comment (kyessenov) [reply source/extensions/listener_managers/listener_manager/lds_api.cc:50]: It's best to split small changes like this in case there are unexpected side-effects.

File: source/common/router/scoped_rds.cc [hunk 2]
```diff
@@ -552,6 +552,22 @@ void ScopedRdsConfigSubscription::onDemandRdsUpdate(
...
+    const envoy::extensions::filters::network::http_connection_manager::v3::ScopedRoutes::
+        ScopeKeyBuilder& scope_key_builder) {
+  applyConfigUpdate([scope_key_builder](ConfigProvider::ConfigConstSharedPtr config)
+                        -> ConfigProvider::ConfigConstSharedPtr {
+    auto* thread_local_scoped_config =
+        const_cast<ScopedConfigImpl*>(static_cast<const ScopedConfigImpl*>(config.get()));
+
+    thread_local_scoped_config->updateScopeKeyBuilder(
+        envoy::extensions::filters::network::http_connection_manager::v3::ScopedRoutes::
+            ScopeKeyBuilder(scope_key_builder));
```
Comment (kyessenov) [inline source/common/router/scoped_rds.cc:565]: Do we need a copy of the proto if it's captured by value?

**Comment by tony612** [comment]
> I'm not super familiar with SRDS, but I have a general question: shouldn't the key be calculated separately from the subscription? I wonder if we have two HCMs pointing to the same SRDS config source with different scope keys, what happens then?

From what I know, I guess the two HCMs will share the same ScopedRdsConfigSubscription. All [scoped route configs](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/scoped_route.proto#envoy-v3-api-msg-config-route-v3-scopedrouteconfiguration) will be used by the two HCMs. But they will use the same scope_key_builder to calculate the scope key, so only one HCM will get the correct result. The difference for the current code and my changed code is which one will get the correct result(earliest or latest updated).

So yes, we may need to separate the key calculation and the subscription. Let me think more about this.

**Comment by tony612** [comment]
@kyessenov Hi, I separated the scope key builder to HCM from SRDS configuration. Please take a look if this is okay.

btw, is there an integration test using two Listeners as an example? So that I can try to add an integration test to test the two HCMs case.

**Comment by kyessenov** [comment]
I looked over the code and it looks like the right approach to me. It would be good to have a test with two listeners with different scope keys if you can add it - not sure if there's one with scoped routes already.

**Comment by tony612** [comment]
@kyessenov OK. It would be better to have an example using two listeners, even without scoped routes. Or I'll find one or try to figure out how to add one.

**Comment by tony612** [comment]
@kyessenov I added an integration test and it's ready to review. Thanks!

**Review by kyessenov — APPROVED** [review]
Thank you, this looks very good!
/assign-from @envoyproxy/senior-maintainers for core HCM change approval.

**Comment by kyessenov** [comment]
Could you update the PR description since the implementation strategy changed?

**Comment by yanavlasov** [comment]
Also assigning @adisuissa to look over the code

File: source/common/http/conn_manager_config.h [hunk 3]
```diff
@@ -336,6 +337,13 @@ class ConnectionManagerConfig {
    */
   virtual Config::ConfigProvider* scopedRouteConfigProvider() PURE;
 
+  /**
+   * @return Router::ScopeKeyBuilder* the scope key builder to calculate the scope key.
+   * Pointer ownership is _not_ transferred to the caller of this function. This will return
+   * nullptr when scoped routing is not enabled.
+   */
+  virtual const Router::ScopeKeyBuilder* scopeKeyBuilder() PURE;
```
Comment (yanavlasov) [inline source/common/http/conn_manager_config.h:345]: From the code it looks like this pointer can never be nullptr. Is it possible to return the reference to make it explicit?

**Review by yanavlasov — COMMENTED** [review]
/wait-any

File: source/common/http/conn_manager_config.h (hunk 3 above)
Comment (tony612) [reply source/common/http/conn_manager_config.h:345]: I think not. It will not be nullptr only when scoped routing is used, just like `scopedRouteConfigProvider`.

**Comment by tony612** [comment]
> Could you update the PR description since the implementation strategy changed?

@kyessenov Done.

File: test/common/router/scoped_rds_test.cc [hunk 4]
```diff
@@ -220,20 +220,26 @@ TEST_F(InlineScopedRoutesTest, InlineRouteConfigurations) {
         key:
           fragments: { string_key: foo-key-2 }
 )EOF");
+  auto config =
```
Comment (adisuissa) [inline test/common/router/scoped_rds_test.cc:223]: nit: const

File: source/common/http/conn_manager_config.h (hunk 3 above)
Comment (adisuissa) [reply source/common/http/conn_manager_config.h:345]: In this case returning OptRef may be better, as it forces the caller to validate that the returned value isn't empty/null.

File: source/common/router/scoped_rds.cc [hunk 5]
```diff
@@ -164,10 +169,8 @@ ScopedRdsConfigSubscription::ScopedRdsConfigSubscription(
             *this, resource_decoder_);
   }
 
-  initialize([scope_key_builder]() -> Envoy::Config::ConfigProvider::ConfigConstSharedPtr {
-    return std::make_shared<ScopedConfigImpl>(
-        envoy::extensions::filters::network::http_connection_manager::v3::ScopedRoutes::
-            ScopeKeyBuilder(scope_key_builder));
+  initialize([]() -> Envoy::Config::ConfigProvider::ConfigConstSharedPtr {
```
Comment (adisuissa) [inline source/common/router/scoped_rds.cc:172]: I'm not sure what's the benefit of having this callback now (no need to access the `scope_key_builder` on initialization of the object).
I don't think that this should be taken care of as part of this PR, but a TODO for investigation can be added here.

File: source/extensions/filters/network/http_connection_manager/config.h [hunk 6]
```diff
@@ -314,6 +316,7 @@ class HttpConnectionManagerConfig : Logger::Loggable<Logger::Id::config>,
   std::chrono::milliseconds request_headers_timeout_;
   Router::RouteConfigProviderSharedPtr route_config_provider_;
   Config::ConfigProviderPtr scoped_routes_config_provider_;
+  Router::ScopeKeyBuilderPtr scope_key_builder_;
```
Comment (adisuissa) [inline source/extensions/filters/network/http_connection_manager/config.h:319]: Can you explain (in a comment) what's the relation between the `scoped_routes_config_provider_` and `scope_key_builder_`? Specifically I wonder if only one can be non-null and the other will.

File: test/integration/scoped_rds_integration_test.cc [hunk 7]
```diff
@@ -564,5 +564,108 @@ route_configuration_name: foo_route1
       /*cluster_0*/ 0);
 }
 
+TEST_P(ScopedRdsIntegrationTest, ListenersWithDifferentScopeKeyBuilder) {
```
Comment (adisuissa) [inline test/integration/scoped_rds_integration_test.cc:567]: Please add a short comment describing what the test validates.

**Review by adisuissa — COMMENTED** [review]
Thanks for detecting the issue and fixing it!
This PR LGTM, although it's hard to understand what was the change, and what was a refactor.
I've got a few minor comments.

File: test/common/http/conn_manager_impl_test_2.cc [hunk 8]
```diff
@@ -2750,6 +2750,10 @@ TEST_F(HttpConnectionManagerImplTest, TestSrdsRouteNotFound) {
   setup(false, "", true, true);
   setupFilterChain(1, 0); // Recreate the chain for second stream.
 
+  // EXPECT_CALL(*scopeKeyBuilder(), computeScopeKey(_)).Times(2);
```
Comment (adisuissa) [inline test/common/http/conn_manager_impl_test_2.cc:2753]: nit: remove comments if not relevant

**Review by adisuissa — COMMENTED** [review]
Thanks!
Left a small nit.
Please also look at CI, seems relevant.

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!
/assign-from @envoyproxy/senior-maintainers

**Review by yanavlasov — APPROVED** [review]
